src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -541,11 +553,23 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
           caller.name, "()");
   }
 
+  // IIS: Ensure that the startup hook is inserted into System.Web.Compilation.BuildManager.InvokePreStartInitMethods.
+  // This will be the first call-site considered for the startup hook injection,
+  // which correctly loads Datadog.Trace.ClrProfiler.Managed.Loader into the application's
+  // own AppDomain because at this point in the code path, the ApplicationImpersonationContext
+  // has been started.
+  auto valid_startup_hook_callsite = true;
+  if (module_metadata->assemblyName == "System"_W ||
+     (module_metadata->assemblyName == "System.Web"_W && !(caller.type.name == "System.Web.Compilation.BuildManager"_W && caller.name == "InvokePreStartInitMethods"_W))) {
+    valid_startup_hook_callsite = false;
+  }
+
   // The first time a method is JIT compiled in an AppDomain, insert our startup
   // hook which, at a minimum, must add an AssemblyResolve event so we can find
   // Datadog.Trace.ClrProfiler.Managed.dll and its dependencies on-disk since it
   // is no longer provided in a NuGet package
-  if (first_jit_compilation_app_domains.find(module_metadata->app_domain_id) ==
+  if (valid_startup_hook_callsite &&
+      first_jit_compilation_app_domains.find(module_metadata->app_domain_id) ==
```

Comment:
Is this saying that we're only going to register a startup hook if the last appdomain loaded is this one?
Does that prevent multiple loads per appdomain?

Reply:
`first_jit_compilation_app_domains.find()` returns the special `end()` iterator if the AppDomain id is _not_ found in the collection. If the following condition is `true`, the value was _not_ found:
```
first_jit_compilation_app_domains.find(module_metadata->app_domain_id) == first_jit_compilation_app_domains.end()
```

[`std::unordered_set<Key,Hash,KeyEqual,Allocator>::find`](https://en.cppreference.com/w/cpp/container/unordered_set/find)

[`std::unordered_set<Key,Hash,KeyEqual,Allocator>::end`](https://en.cppreference.com/w/cpp/container/unordered_set/end)

The goal is to inject the startup hook only once per AppDomain, so we track all the ids we already injected into.